File: swiftui/CellposeGUI/Services/SidecarProcessManager.swift
import Foundation
import Observation

@MainActor
@Observable
final class SidecarProcessManager {
    private(set) var isReady = false
    private(set) var hasFailed = false
    private(set) var statusMessage = "Starting sidecar…"
    private(set) var baseURL: URL

    private var process: Process?
    private var repoRoot: URL
    private var stderrBuffer = ""

    init(baseURL: URL? = nil) {
        if let baseURL {
            self.baseURL = baseURL
            self.repoRoot = SidecarProcessManager.defaultRepoRoot()
        } else if let envURL = ProcessInfo.processInfo.environment["SIDECAR_URL"],
                  let url = URL(string: envURL)
        {
            self.baseURL = url
            self.repoRoot = SidecarProcessManager.defaultRepoRoot()
        } else {
            self.baseURL = URL(string: "http://127.0.0.1:8787")!
            self.repoRoot = SidecarProcessManager.defaultRepoRoot()
        }
    }

    static func defaultRepoRoot() -> URL {
        if let envRoot = ProcessInfo.processInfo.environment["CELLPOSE_ROOT"],
           !envRoot.isEmpty
        {
            return URL(fileURLWithPath: envRoot, isDirectory: true)
        }

        for start in [
            Bundle.main.bundleURL,
            URL(fileURLWithPath: FileManager.default.currentDirectoryPath, isDirectory: true),
        ] {
            if let root = findRepoRoot(startingAt: start) {
                return root
            }
        }

        return URL(fileURLWithPath: FileManager.default.currentDirectoryPath, isDirectory: true)
    }

    private static func findRepoRoot(startingAt url: URL) -> URL? {
        var current = url.standardizedFileURL
        let fileManager = FileManager.default

        while !current.path.isEmpty, current.path != "/" {
            let cellposeDir = current.appendingPathComponent("cellpose")
            let pyproject = current.appendingPathComponent("pyproject.toml")
            if fileManager.fileExists(atPath: cellposeDir.path),
               fileManager.fileExists(atPath: pyproject.path)
            {
                return current
            }
            current = current.deletingLastPathComponent()
        }

        return nil
    }

    private static func findUvExecutable() -> String {
        let pathEntries = (ProcessInfo.processInfo.environment["PATH"] ?? "")
            .split(separator: ":", omittingEmptySubsequences: true)
            .map(String.init)

        for entry in pathEntries {
            let candidate = URL(fileURLWithPath: entry).appendingPathComponent("uv").path
            if FileManager.default.isExecutableFile(atPath: candidate) {
                return candidate
            }
        }

        let home = FileManager.default.homeDirectoryForCurrentUser.path
        for candidate in [
            "/opt/homebrew/bin/uv",
            "/usr/local/bin/uv",
            "\(home)/.local/bin/uv",
            "\(home)/.cargo/bin/uv",
        ] {
            if FileManager.default.isExecutableFile(atPath: candidate) {
                return candidate
            }
        }

        return "uv"
    }

    func startIfNeeded() async {
        if RuntimeEnvironment.skipSidecar {
            isReady = true
            statusMessage = "Sidecar skipped"
            return
        }

        if await probeHealth() {
            isReady = true
            statusMessage = "Sidecar ready"
            return
        }

        guard process == nil else { return }
        statusMessage = "Launching Python sidecar…"

        let uvPath = Self.findUvExecutable()
        let proc = Process()
        proc.currentDirectoryURL = repoRoot
        proc.executableURL = URL(fileURLWithPath: uvPath)
        proc.arguments = [
            "run", "python", "-m", "cellpose.api",
            "--host", "127.0.0.1",
            "--port", "\(baseURL.port ?? 8787)",
        ]

        let stderrPipe = Pipe()
        proc.standardOutput = Pipe()
        proc.standardError = stderrPipe

        stderrBuffer = ""
        stderrPipe.fileHandleForReading.readabilityHandler = { [weak self] handle in
            let data = handle.availableData
            guard !data.isEmpty, let text = String(data: data, encoding: .utf8) else { return }
            Task { @MainActor in
                self?.stderrBuffer += text
            }
        }

        do {
            try proc.run()
            process = proc
        } catch {
            hasFailed = true
            statusMessage = "Failed to launch sidecar: \(error.localizedDescription)"
            return
        }

        // PyTorch + first import can take well over 30s on cold start.
        for attempt in 0 ..< 180 {
            if let process, !process.isRunning {
                let exitCode = process.terminationStatus
                stderrPipe.fileHandleForReading.readabilityHandler = nil
                let detail = stderrBuffer.trimmingCharacters(in: .whitespacesAndNewlines)
                hasFailed = true
                statusMessage = detail.isEmpty
                    ? "Sidecar exited with code \(exitCode)."
                    : "Sidecar failed: \(detail)"
                self.process = nil
                return
            }

            if await probeHealth() {
                stderrPipe.fileHandleForReading.readabilityHandler = nil
                isReady = true
                statusMessage = "Sidecar ready"
                return
            }

            statusMessage = "Waiting for sidecar (\(attempt + 1)/180)…"
            try? await Task.sleep(nanoseconds: 1_000_000_000)
        }

        stderrPipe.fileHandleForReading.readabilityHandler = nil
        let detail = stderrBuffer.trimmingCharacters(in: .whitespacesAndNewlines)
        hasFailed = true
        statusMessage = detail.isEmpty
            ? "Sidecar failed to start. Run `uv run python -m cellpose.api` manually to see errors."
            : "Sidecar failed to start: \(detail)"
    }

    func stop() {
        process?.terminate()
        process = nil
        isReady = false
    }

    private func probeHealth() async -> Bool {
        let client = SidecarClient(baseURL: baseURL)
        do {
            let health = try await client.get("/health", as: SidecarHealth.self)
            return health.status == "ok"
        } catch {
            return false
        }
    }
}

File: winui/CellposeGUI/Services/SidecarProcessManager.cs
using System.Diagnostics;
using System.Text;
using CellposeGUI.Models;

namespace CellposeGUI.Services;

public sealed class SidecarProcessManager
{
    private Process? _process;
    private readonly StringBuilder _stderr = new();

    public Uri BaseUri { get; }
    public string RepoRoot { get; }
    public bool IsReady { get; private set; }
    public string StatusMessage { get; private set; } = "Starting sidecar…";

    public event Action<string>? StatusChanged;

    public SidecarProcessManager(Uri? baseUri = null)
    {
        var envUrl = Environment.GetEnvironmentVariable("SIDECAR_URL");
        if (baseUri != null)
            BaseUri = baseUri;
        else if (!string.IsNullOrWhiteSpace(envUrl) && Uri.TryCreate(envUrl, UriKind.Absolute, out var parsed))
            BaseUri = parsed;
        else
            BaseUri = new Uri("http://127.0.0.1:8787");

        RepoRoot = ResolveRepoRoot();
    }

    public async Task StartIfNeededAsync(CancellationToken cancellationToken = default)
    {
        SetStatus("Checking sidecar…");
        if (await ProbeHealthAsync(cancellationToken))
        {
            IsReady = true;
            SetStatus("Sidecar ready");
            return;
        }

        if (_process != null)
            return;

        SetStatus("Launching Python sidecar…");
        var port = BaseUri.IsDefaultPort ? 8787 : BaseUri.Port;
        var uvPath = FindUvExecutable();

        _process = new Process
        {
            StartInfo = new ProcessStartInfo
            {
                FileName = uvPath,
                Arguments = $"run python -m cellpose.api --host 127.0.0.1 --port {port}",
                WorkingDirectory = RepoRoot,
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
            },
            EnableRaisingEvents = true,
        };

        _process.ErrorDataReceived += (_, e) =>
        {
            if (!string.IsNullOrWhiteSpace(e.Data))
                _stderr.AppendLine(e.Data);
        };

        try
        {
            if (!_process.Start())
            {
                SetStatus("Failed to launch sidecar process.");
                _process = null;
                return;
            }

            _process.BeginErrorReadLine();
        }
        catch (Exception ex)
        {
            SetStatus($"Failed to launch sidecar: {ex.Message}");
            _process = null;
            return;
        }

        // PyTorch + first import can take well over 30s on cold start.
        for (var attempt = 0; attempt < 180; attempt++)
        {
            cancellationToken.ThrowIfCancellationRequested();

            if (_process.HasExited && _process.ExitCode != 0)
            {
                var detail = _stderr.ToString().Trim();
                SetStatus(string.IsNullOrWhiteSpace(detail)
                    ? $"Sidecar exited with code {_process.ExitCode}."
                    : $"Sidecar failed: {detail}");
                _process = null;
                return;
            }

            if (await ProbeHealthAsync(cancellationToken))
            {
                IsReady = true;
                SetStatus("Sidecar ready");
                return;
            }

            SetStatus($"Waiting for sidecar ({attempt + 1}/180)…");
            await Task.Delay(1000, cancellationToken);
        }

        var stderrTail = _stderr.ToString().Trim();
        SetStatus(string.IsNullOrWhiteSpace(stderrTail)
            ? "Sidecar failed to start. Run `uv run python -m cellpose.api` manually to see errors."
            : $"Sidecar failed to start: {stderrTail}");
    }

    public void Stop()
    {
        if (_process == null)
            return;

        try
        {
            if (!_process.HasExited)
                _process.Kill(entireProcessTree: true);
        }
        catch
        {
            // Best effort shutdown.
        }
        finally
        {
            _process.Dispose();
            _process = null;
            IsReady = false;
        }
    }

    private void SetStatus(string message)
    {
        StatusMessage = message;
        StatusChanged?.Invoke(message);
    }

    private async Task<bool> ProbeHealthAsync(CancellationToken cancellationToken)
    {
        try
        {
            var client = new SidecarClient(BaseUri);
            var health = await client.GetAsync<SidecarHealth>("/health", cancellationToken);
            return health.Status == "ok";
        }
        catch
        {
            return false;
        }
    }

    private static string ResolveRepoRoot()
    {
        var envRoot = Environment.GetEnvironmentVariable("CELLPOSE_ROOT");
        if (!string.IsNullOrWhiteSpace(envRoot) && Directory.Exists(envRoot))
            return Path.GetFullPath(envRoot);

        foreach (var start in new[]
                 {
                     Directory.GetCurrentDirectory(),
                     AppContext.BaseDirectory,
                 })
        {
            var found = FindRepoRoot(start);
            if (found != null)
                return found;
        }

        return Directory.GetCurrentDirectory();
    }

    private static string? FindRepoRoot(string startPath)
    {
        var current = new DirectoryInfo(Path.GetFullPath(startPath));
        while (current != null)
        {
            if (Directory.Exists(Path.Combine(current.FullName, "cellpose")) &&
                File.Exists(Path.Combine(current.FullName, "pyproject.toml")))
                return current.FullName;

            current = current.Parent;
        }

        return null;
    }

    private static string FindUvExecutable()
    {
        var pathEntries = (Environment.GetEnvironmentVariable("PATH") ?? "")
            .Split(Path.PathSeparator, StringSplitOptions.RemoveEmptyEntries);

        foreach (var entry in pathEntries)
        {
            var candidate = Path.Combine(entry, OperatingSystem.IsWindows() ? "uv.exe" : "uv");
            if (File.Exists(candidate))
                return candidate;
        }

        return "uv";
    }
}

